teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+
+- Achieve SLSA L3 supply chain security through Tekton `PipelineRuns`
+- End-to-end trust of the provenance produced by Tekton Chains for Tekton `PipelineRuns`
+- `Tasks` to emit a list of artifacts of unknown length.
+
+### Requirements
+
+- Enable `Tasks` to be able to surface Artifact metadata as an array.
+
+## Proposal
+
+Define the structure of the Artifact metadata that the `Step` needs to write out. Define the file path and variable replacement syntax for the location of the metadata file.
+
+The following chapters in the proposal describe in details the proposed phased implementation approach:
+
+- [Artifact Provenance Data](#artifact-metadata)
+- [surfacing artifact metadata into the status](#from-step-into-status)
+- [surfacing the artifact metadata into the provenance payload](#from-status-into-provenance)
+
+
+### Artifact Provenance Data
+
+An artifact can be consumed or produced by a `Task`. This means that we will always have two categories at a high level: `inputs` (to indicate what was downloaded) and `outputs` (to indicate what was uploaded). The metadata content could be defined as follows:
+
+```json
+{
+    "inputs":[
+        "name": "artifact Name", # e.g. source code, dependencies
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ],
+    "outputs": [
+        "name": "artifact Name", # e.g. images, packages, coverage reports
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ]
+}
+```
+The metadata content is written by the `Step` to a file `$(step.artifacts.path)`:
+
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee $(step.artifacts.path)
+```
+The path `$(step.artifacts.path)` resolves to `/tekton/steps/<step-name>/artifacts/provenance.json` in the `Step` container at runtime.
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
+Name is a `Task`/`Step` defined name. It is useful for referencing (phase 2) and UI. The name the artifact under a particular category (`input` or `output`) must be unique.
```

Comment:
If this is specific to phase 2, should we omit it for now?

Reply:
I was thinking that it would also be a nice addition for things like dashboard or other tools where it would be nice to distinguish between the different input and output artifacts. e.g we may want to distinguish between different packages vs images produced by the build.

For phase2, I was thinking that the name here should match the name of the artifact defined in the API spec. This way, its an easy transition and the internals of a Step still remain the same.

Reply:
Made it optional like we discussed in the Data Interfaces WG meeting today.